[spmd] Introduce ParallelMode and add DTensorExpandMode

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99374
* #99489
* #99436
* #99373
* #99435
* #99062
* __->__ #98452

This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

Put DTensorExpandMode as the first parallel mode that does the
existing dtensor_expand functionality.

Differential Revision: [D45174399](https://our.internmc.facebook.com/intern/diff/D45174399)

wanchaol committed:
[spmd] add parallel mode

This PR adds a paralell mode interface to define how to expand
and optimize the graphs. This would be beneifical for ddp/fsdp
to expand differently and apply different optimization passes

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -393,6 +340,7 @@ class _CompiledResult:
 def _compile(
     func: Callable,
     module_override: Optional[Dict[Type[Any], Override]],
+    parallel_mode: Optional[ParallelMode],
```
Does this mean, we will have dedicated DDP and FSDp parallel mode which can be plugged in independently?

wanchaol replied:
Yeah, I would like to add a dedicated DataParallelMode later where it consists of 3 styles as we described in the design (default, replicate, fully_shard). then base on the style the expansion would be different and would generate its corresponding pattern. I am not sure if we want to shard the parameters/states in the `parallel_mode.expand` api directly, so looking for some suggestions :)

fegin replied:
Consolidating all the information into once place is good. However, this can also pose an issue that there will be combinatorial default modes we will have to set (e.g., DDP, FSDP, TP, SPARSE and they can be combined). Additionally, if users only want to override the transformation part (users may not have good knowledge or don't care about expansion), this seems to be a burden.

wanchaol replied:
Oh I think for composable parallelism (i.e. DDP/FSDP/TP/Sparse/Pipeline), we should either create a "manual" parallel mode that manually composes all different expansion and optimizations, or we should rely on the "auto" parallel mode directly? It's a hard to answer question before we really hit such a case though..

mrshenli commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,101 @@
+    def expand(
+        self,
+        gm: GraphModule,
+        params_and_buffers: Dict[str, Any],
```
will this support hybrid paradigms? I assume yep, we basically just need to tag the params and buffers accordingly and then users can plugin a parallel mode implementation they prefer for that?

wanchaol replied:
My view on this: if user don't tag using composable API, we by default use a default mixed sharding. If user tag with composable, then in the composable we set the DataParallelMode to its corresponding style (replicate or fully shard), the expand code would expand correspondingly

fegin commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,101 @@
+
+        return _convert_to_distributed(gm, inps, schemas, _allow_partial=False)
+
+    def optimize(self, gm: GraphModule) -> GraphModule:
```
Prefer to use `transform` name. Some graph transformation that is not related to performance may also need to be done here. It is not necessary the optimization passes.

wanchaol replied:
I'm taking the term from `dynamo.optimize` and apply similar pattern here, `dynamo.optimize` do both graph capture and lower to inductor for "real compilation". 

I think after SPMD expansion, it's a "functional" distributed graph, anything graph transformation after the SPMD expansion should be related to optimization? Could you share what potential graph transformations not related to optimization?

wanchaol replied:
updated with `transform_and_compile`, let me know if this looks good or not :)

fegin commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,101 @@
+
+    def __init__(self):
+        self._placements_override: Dict[int, List[Placement]] = {}
+        self._optimization_passes: List[Callable] = []
```
Why not just take a callable and remove `optimize` method? This way it is more flexible for users to configure while not losing any connivence provided here.

wanchaol replied:
I think we need to hold a default set of optimization passes. where we explicitly register in this `__init__`. `optimize` calls the optimization passes in the order of this list one by one. Do you mean we have an optional callable argument in the init method, if not passed in, we use the default set of optimization, if passed in, we drop the default set and use the passed in callable?

That works for me :) We could get rid of `configure_optimization_passes` to favor of this!

The `optimize` method is merely a convenient callsite for the `compile` API to run the list of default optimization passes and lower to inductor to do "real" compilation, we can remove it for sure, but we would put this logic to the `compile` API itself.

wanchaol replied:
updated with suggestions

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorFallbackMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorFallbackMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorFallbackMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorFallbackMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorFallbackMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

This PR also intentionally to make the dtensor expand to a separate mode
and try to avoid breaking the existing code path

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorExpandMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

Put DTensorExpandMode as the first parallel mode that does the
existing dtensor_expand functionality.

[ghstack-poisoned]

awgu commented on torch/distributed/_spmd/api.py:
```diff
@@ -567,6 +522,10 @@ def compile(
             compiled and distributed (usually after the first iteration) to
             transform the compiled GraphModule into a new optimized one.
+        parallel_mode (Optional[ParallelMode]): a :class:`ParallelMode` object
+            that specifies how to parallel the callable. Each ParallelMode would
```
nit: maybe?
```suggestion
            that specifies how to parallelize the callable. Each ParallelMode would
```

awgu commented on torch/distributed/_spmd/api.py:
```diff
@@ -567,6 +522,10 @@ def compile(
             transform the compiled GraphModule into a new optimized one.
+        parallel_mode (Optional[ParallelMode]): a :class:`ParallelMode` object
+            that specifies how to parallel the callable. Each ParallelMode would
+            has its own strategy to partition the model and the captured graph
```
nit:
```suggestion
            have its own strategy to partition the model and the captured graph
```

awgu commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,131 @@
+    ) -> GraphModule:
+        flat_args, _ = pytree.tree_flatten(list(args) + list(kwargs.values()))
+
+        mesh = DeviceMesh("cuda", torch.arange(dist.get_world_size()).cuda())
```
nit:
```suggestion
        mesh = DeviceMesh("cuda", torch.arange(dist.get_world_size(), device="cuda"))
```

constroy replied:
Wonder why the mesh Tensor is on `cuda` device?

awgu commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,131 @@
+
+        inps, schemas = [], []
+
+        for p in pytree.tree_flatten(params_and_buffers)[0]:
```
I have not been following the code closely, but I noticed this `pytree.tree_flatten()` a lot in PyTorch compiler code. I was curious what properties it has such that you want to use it here instead of something like:
```
next(iter(params_and_buffers))
```

Also, this seems to be an unprotected access and will error if `params_and_buffers` is empty? Same with `named_states` below.

awgu commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,131 @@
+
+        with FakeTensorMode(allow_non_fake_inputs=True):
+            fake_inps = [torch.empty_like(inp) for inp in inps]
+
```
nit: Do you think it is worth for readability to assert or document that the length of `schemas` and `fake_inputs`/`inps` is `len(params_and_buffers) + len(named_states) + len(args) + len(kwargs)`, or is that already clear?

wanchaol replied:
It's not really the same, in train_step, we have `mod` and `opt` in the arguments too, so they might be slightly different sometimes as we can't control what things user pass to the train step

awgu reviewed (commented):
Reading some code... :)

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorExpandMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

Put DTensorExpandMode as the first parallel mode that does the
existing dtensor_expand functionality.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,131 @@
+        # FIXME: allow other sharding schemas
+        replicate_schema: Schema = Schema(mesh=mesh, placements=[Replicate()])
+
+        inps, schemas = [], []
```
I am basically copy-pasta existing code from the other file so didn't want to change the code in this PR. I can submit new PRs to fix those type annotations :)

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorExpandMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

Put DTensorExpandMode as the first parallel mode that does the
existing dtensor_expand functionality.

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -179,65 +175,14 @@ def _to_caller_flattened_graph_module(gm: torch.fx.GraphModule) -> torch.fx.Grap
 
-_placements_override: Dict[int, List[Placement]] = {}
+# Use a dtensor expand mode for now to preserve the old behavior
+# and avoid break existing code
```
nit: avoid breaking

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -578,7 +537,12 @@ def wrapper(*args, **kwargs):
             if compiled_obj is None:
                 first_iter = True
-                compiled_obj = _compile(func, module_override, *args, **kwargs)
+                global dtensor_expand_mode
```
(no need to change this PR) curious why do create `dtensor_expand_mode` as a global instead of creating a new one for every recompilation. Do we expect we need pass states across recompilations or will the constructor becoming more expensive later on?

wanchaol replied:
The main reason is that there's a `_override_placements` API that uses `_placements_override` (i think it's for sparse part integration), where `DTensorExpandMode` put this as an attribute, so it need to be a global var now in order for the `_override_placements` to still work, we might can put this into `DTensorExpand` mode too but that would also be BC breaking.

mrshenli reviewed (approved):
LGTM, left some minor comments

wanchaol commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -0,0 +1,131 @@
+    def partition(
+        self,
+        gm: GraphModule,
+        model: torch.nn.Module,
+        optimizer: Optional[torch.optim.Optimizer],
```
Yes it would be needed in PRs on top

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorExpandMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

Put DTensorExpandMode as the first parallel mode that does the
existing dtensor_expand functionality.

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorExpandMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

Put DTensorExpandMode as the first parallel mode that does the
existing dtensor_expand functionality.

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce ParallelMode and add DTensorExpandMode"


This PR introduces a ParallelMode interface to define how to do
SPMD expansion and optimize the captured graph. This would be
beneifical for different parallelisms to expand differently
and apply different optimization passes

Put DTensorExpandMode as the first parallel mode that does the
existing dtensor_expand functionality.

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45174399).

wanchaol commented:
@pytorchbot merge
